lib/srv/ctx.go

```diff
@@ -678,16 +680,66 @@ func (c *ServerContext) ExecCommand() (*ExecCommand, error) {
 	var pamEnabled bool
 	var pamServiceName string
 	var pamUseAuth bool
+	var pamEnvironmentRaw map[string]string
+	var pamEnvironment map[string]string = make(map[string]string)
 
 	// If this code is running on a node, check if PAM is enabled or not.
+	// Override cluster configuration with optional local configuration.
 	if c.srv.Component() == teleport.ComponentNode {
-		conf, err := c.srv.GetPAM()
+		accessPoint := c.srv.GetAccessPoint()
+		clusterPAMConfig, _ := accessPoint.GetPAMConfig()
+		localPAMConfig, err := c.srv.GetPAM()
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-		pamEnabled = conf.Enabled
-		pamServiceName = conf.ServiceName
-		pamUseAuth = conf.UsePAMAuth
+
+		if clusterPAMConfig == nil || localPAMConfig.Enabled {
```

Comment:
Factor pam configuration query into a separate function for readability:
```go
var pamConfig pamConfig
// If this code is running on a node, check if PAM is enabled or not.
// Override cluster configuration with optional local configuration.
if c.srv.Component() == teleport.ComponentNode {
	config, err := c.getPAMConfig()
	if err != nil {
		return nil, trace.Wrap(err)
	}
	pamConfig = *config
}
// ...
```
and `getPAMConfig`:
```go
func (c *ServerContext) getPAMConfig() (config *pamConfig, err error) {
	accessPoint := c.srv.GetAccessPoint()
	clusterPAMConfig, err := accessPoint.GetPAMConfig()
	if err != nil {
		return nil, trace.Wrap(err)
	}
	localPAMConfig, err := c.srv.GetPAM()
	if err != nil {
		return nil, trace.Wrap(err)
	}

	var environ map[string]string
	// Use local configuration by default
	config = &pamConfig{
		enabled:     localPAMConfig.Enabled,
		serviceName: localPAMConfig.ServiceName,
		useAuth:     localPAMConfig.UsePAMAuth,
		environ:     make(map[string]string),
	}
	// But override with cluster configuration if available
	if clusterPAMConfig != nil {
		config.enabled = clusterPAMConfig.GetEnabled()
		config.serviceName = clusterPAMConfig.GetServiceName()
		config.useAuth = clusterPAMConfig.GetUsePAMAuth()
		environ = clusterPAMConfig.GetEnvironment()
	}

	user, err := accessPoint.GetUser(c.Identity.TeleportUser, false)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	traits := user.GetTraits()

	// Interpolate any expressions in the cluster PAM environment configuration
	// with values from received traits.
	for key, value := range environ {
		expr, err := parse.NewExpression(value)
		if err != nil {
			return nil, trace.Wrap(err)
		}

		if expr.Namespace() != "external" {
			return nil, trace.BadParameter("only variables in external namespace allowed in PAM configuration")
		}

		result, err := expr.Interpolate(traits)
		if err != nil {
			return nil, trace.Wrap(err)
		}

		config.environ[key] = strings.Join(result, " ")
	}
	return config, nil
}

type pamConfig struct {
	enabled     bool
	serviceName string
	useAuth     bool
	environ     map[string]string
}
```

Reply:
adjusted